In [1]:
import numpy as np
import matplotlib.pyplot as plt
import sympy as sy
import utils as utils
from fractions import Fraction
from func import Func
import func

from IPython.display import display, HTML

# Inline plotting
%matplotlib inline

# Make sympy print pretty math expressions
sy.init_printing()

utils.load_custom_styles()

---
# Exercise 1



<img src="figures/exam-jun-2018/ex1.png" width="800" />



---
## Exercise 1 a)

<img src="figures/exam-jun-2018/ex1a.png" width="800" />





<img src="figures/exam-jun-2018/sketch.jpg" width="500" />














---
## Exercise 1 b)

<img src="figures/exam-jun-2018/ex1b.png" width="800" />





Using following procedure:

**Step 1:** Convert inequalities to equalites using slack variables

**Step 2:** Convert objective function to an equally.

**Step 3:** Construct the initial simplex tableaux


<img src="figures/exam-jun-2018/initial-tableaux.jpg" width="700" />





**Step 4:** Find the pivot column by looking at the smallest value at the bottom row. 

**Step 5:** Find the pivot row by pick the smallest of the ratios: $b_i / a_{i1}$

**Step 6:** Make the pivot column to a basis column using elementary row operations.

Following row operations are needed:

- $\frac{1}{2} R_3  \to R_3$
- $R_1  - R_3 \to R_1$
- $R_2  - 6 R_3 \to R_2$
- $R_4  + 6 R_3 \to R_4$


In [2]:
from simplex import SimplexSolver 
A = [[1, 4],
     [6, 4],
     [2, -5]]
b = [16, 30, 6]
c = [6, 5]
solver = SimplexSolver()
res = solver.run_simplex(A, b, c, prob='max', enable_msg=True, latex_path=None)

   x_1   x_2   s_1   s_2   s_3    b    
|   1     4     1     0     0    16   | s_1
|   6     4     0     1     0    30   | s_2
|   2    -5     0     0     1     6   | s_3
|  -6    -5     0     0     0     0   |
Current solution: {'x_1': 0, 'x_2': 0, 's_1': Fraction(16, 1), 's_2': Fraction(30, 1), 's_3': Fraction(6, 1), 'z': 0}

   x_1   x_2   s_1   s_2   s_3    b    
|   1     4     1     0     0    16   | s_1
|   6     4     0     1     0    30   | s_2
|   2    -5     0     0     1     6   | s_3
|  -6    -5     0     0     0     0   |

There are negative elements in the bottom row, so the current solution is not optimal. Thus, pivot to improve the current solution. The entering variable is x_1 and the departing variable is s_3.


Perform elementary row operations until the pivot is one and all other elements in the entering column are zero.

   x_1   x_2   s_1   s_2   s_3    b    
|   0   13/2    1     0   -1/2   13   | s_1
|   0    19     0     1    -3    12   | s_2
|   1   -5/2    

---
## Exercise 1c)

<img src="figures/exam-jun-2018/ex1c.png" width="800" />





We know that at least one of the extreme points or corners of the feasible area corresponds to an optimal solution:

<img src="figures/exam-jan-2018/theorem-6.png" width="800" />





Therefore, we only need to evaluate the corner points that have integer entries. There are only three such points: 
- $f(0,0)= 6 \cdot 0 + 5 \cdot 0 = 0$
- $f(0,4)= 6 \cdot 0 + 5 \cdot 4 = 20$
- $f(0,4)= 6 \cdot 0 + 5 \cdot 3 = 15$

Thus, $(0,4)$ is the maximum integer point.

---
## Exercise 2 a)

<img src="figures/exam-jun-2018/ex2a.png" width="800" />





The gradient of $f$:
$$
\nabla f =  \begin{bmatrix} \frac{\partial f}{\partial x_1} \\ \frac{\partial f}{\partial x_2} \\ \end{bmatrix}= \begin{bmatrix} x_2 - 1 \\ x_1 \\  \end{bmatrix}
$$

The gradient of $f$ gives the partial derivatives in each dimensions. The partial derivative $\frac{\partial f}{\partial x_1}$ gives us the slope of $f$ in the positive $x_1$ direction and $\frac{\partial f}{\partial x_2}$ gives us the slope of $f$ in the positive $x_2$ direction.

Instead of fixing the direction, we can generalise the partial derivatives to compute the slope in any direction that we want. Like the partial derivative, we can define the directional derivative as follows:

\begin{align*}
D_{\mathbf{d}} f(\mathbf{x}) = \lim_{h \to 0} \frac{f(\mathbf{x} + h \mathbf{d}) - f(\mathbf{x})}{h}
\end{align*}

This means that we can dot the gradient with the direction vector $\mathbf{d}$ to find the directional derivative:

The directional derivative:
$$
\frac{\partial f(\mathbf{x})}{\partial \mathbf{d}}  = \nabla f(\mathbf{x})^T \mathbf{d} = [x_2 -1, x_1] \begin{bmatrix} 1 \\ 1 \\ \end{bmatrix} = x_1 + x_2 - 1
$$


In [13]:
x1, x2 = sy.symbols('x1, x2')
f = sy.Matrix([x1*x2 - x1])
d = f.jacobian([x1, x2]).T
d

⎡x₂ - 1⎤
⎢      ⎥
⎣  x₁  ⎦

In [12]:
d.T * sy.Matrix([[1], [1]])

[x₁ + x₂ - 1]

---
## Exercise 2 b)

<img src="figures/exam-jun-2018/ex2b.png" width="800" />





Since it is an optimisation problem with equality constraints, we can use the Lagrangian method.

The Lagrangian method is given as follows:
1. Define the Lagrangian function by incoporating the constraint into the objective function
$$
L(\mathbf{x}, \lambda) = f(\mathbf{x}) - \lambda ( h(\mathbf{x}) - c  )
$$
where $\lambda$ is the Lagrangian multiplier associated with the contraint. When the contraint holds i.e., when $h(\mathbf{x})=c$ then $L(\mathbf{x}, \lambda)=f(\mathbf{x})$
2. Since $L(\mathbf{x}, \lambda)$ is an unconstraint objective function, we can optimise it by finding its FONC i.e., taking the derivative and setting it zero.
3. Now, we get a system of linear equations. Solving it will yield the points where the objective function is maximised or minimised.

In [14]:
x1, x2 = sy.symbols('x1, x2')
f = x1*x2 - x1
h = x1**2 + x2**2 - 4
F = Func(f, (x1, x2), constraints=[h])
F

[x₁⋅x₂ - x₁]

In [41]:
points, lambdas = F.solve_lagrangian()

In [40]:
dec_points = []
for point in points:
    l = [float(i) for i in list(point)]
    dec_points.append(l)
F.evalf(dec_points)

Evaluting f[1.0756066517006793, 1.6861406616345072] = 0.738017459656381
Evaluting f[-1.6103013167780909, -1.1861406616345072] = 3.52034518609217
Evaluting f[1.6103013167780909, -1.1861406616345072] = -3.52034518609217
Evaluting f[-1.0756066517006793, 1.6861406616345072] = -0.738017459656381


Maximum is $[-1.6103013167780909, -1.1861406616345072]^T$ and minimum is $[1.6103013167780909, -1.1861406616345072]^T$.

---
## Exercise 2 c)

<img src="figures/exam-jun-2018/ex2c.png" width="800" />





The new constraint $D$ turns the problem into a canonical linear programming problem. For this kind of problems, we can use Theorem 6:

<img src="figures/exam-jan-2018/theorem-6.png" width="800" />

Thus, the maximum of $f$ lies on the circle which we found in b).






---
# Exercise 3

<img src="figures/exam-jun-2018/ex3.png" width="800" />





---
## Exercise 3 a)

<img src="figures/exam-jun-2018/ex3a.png" width="800" />





No.

---
## Exercise 3 b)

<img src="figures/exam-jun-2018/ex3b.png" width="800" />





Continuous optimisation is when the input to our objective function is continuous.

---
## Exercise 3 c)

<img src="figures/exam-jun-2018/ex3c.png" width="800" />





Stochastic methods such as genetic algorithm employ randomness in the search for an optimum.

---
## Exercise 3 d)

<img src="figures/exam-jun-2018/ex3d.png" width="800" />




It means that these methods use a set of candidate solutions to find the best solution. One example of such an optimisation method is particle swarm.

---
## Exercise 3 e)

<img src="figures/exam-jun-2018/ex3e.png" width="800" />






### Exercise 3 e. i)

<img src="figures/exam-jun-2018/ex3i.png" width="800" />





A particle can be represented as a vector with the following information:
- position, two values $(x_p,y_o)$
- velocity, two values $(x_v,y_v)$
- fitness value, evaluate the function at the current position: $f(x_p, y_o)$
- best position to keep track of the particle's personal best

---
## Exercise 3 ii)

<img src="figures/exam-jun-2018/ex3j.png" width="800" />





It refers to the velocity for particle $i$.

---
## Exercise 3 iii)

<img src="figures/exam-jun-2018/ex3k.png" width="800" />





It refers to iteration of the iterative algorithm.

---
## Exercise 3 iv)

<img src="figures/exam-jun-2018/ex3l.png" width="800" />





- Step 1: Compute the fitness of each particle
- Step 2: Update the best position of each particle and the best global particle
- Step 3: Update each individual particle's velocity and position

---
# Exercise 4, Question 1


<img src="figures/exam-jun-2018/ex4.png" width="800" />





---
## Exercise 4, Question 1 a)


<img src="figures/exam-jun-2018/ex4.1a.png" width="800" />





We can use Baye's decision rule to classify the test samples:

- Classify $x$ as $c_1$ if $P(c_1 \mid x) > P(c_2 \mid x)$; otherwise classify $x$ as $c_2$

Using Bayes' formula, we have:

$$
P(c_k \mid \mathbf{x}) = \frac{p(\mathbf{x} \mid c_k) P(c_k)}{p(\mathbf{x})}
$$
where
$$
p(\mathbf{x}) = \sum_{k=1}^K p(\mathbf{x} \mid c_k) P(c_k)
$$

Since $p(\mathbf{x})$ is the same for when computing $P(c_k\mid \mathbf{x})$, it is not necessary to compute it when we want to use the conditional probability to classify new samples. Therefore, the decision rule becomes:

- Classify $x$ as $c_1$ if $p(x \mid c_1)P(c_1) \geq p(x \mid c_2) P (c_2)$; otherwise classify $x$ as $c_2$

The prior probabilities $P(c_1)$ and $P(c_2)$ are given:
- $P(c_1) = 100/300=1/3$
- $P(c_2) = 200/300=2/3$


We can now classify:

$x_1$ is classified as $c_1$ because $p(x = 3 \mid c_1) P(c_1) > p(x = 3 \mid c_2) P(c_2)$:
$$
p(x = 3 \mid c_1) P(c_1) = \frac{1}{10} \cdot \frac{1}{3} = \frac{1}{30} \\
p(x = 3 \mid c_2) P(c_2) = 0 \cdot \frac{2}{3} = 0
$$

$x_2$ is classified as $c_1$ because $p(x = 7 \mid c_1) P(c_1) = p(x = 7 \mid c_2) P(c_2)$:
$$
p(x = 7 \mid c_1) P(c_1) = \frac{1}{10} \cdot \frac{1}{3} = \frac{1}{30}\\
p(x = 7 \mid c_2) P(c_2) = \frac{1}{20} \cdot \frac{2}{3} = \frac{1}{30}
$$

$x_3$ is classified as $c_2$ because $p(x = 8 \mid c_1) P(c_1) < p(x = 8 \mid c_2) P(c_2)$:
$$
p(x = 8 \mid c_1) P(c_1) = \frac{1}{20} \cdot \frac{1}{3} = \frac{1}{60}\\
p(x = 8 \mid c_2) P(c_2) = \frac{1}{10} \cdot \frac{2}{3} = \frac{1}{15}
$$

$x_4$ is classified as $c_2$ because $p(x = 9 \mid c_1) P(c_1) < p(x = 9 \mid c_2) P(c_2)$:
$$
p(x = 9 \mid c_1) P(c_1) = 0 \cdot \frac{1}{3} = 0\\
p(x = 9 \mid c_2) P(c_2) = \frac{1}{5} \cdot \frac{2}{3} = \frac{2}{15}
$$

---
## Exercise 4, Question 1 b)


<img src="figures/exam-jun-2018/ex4.1b.png" width="800" />





The risk of taking action $\alpha_i$ given the observation $x$ is given as:

$$
R(\alpha_i \mid x) = \lambda(\alpha_i \mid c_1) P(c_1 \mid x) + \lambda(\alpha_i \mid c_2) P(c_2 \mid x)
$$

Classify $x_1$ as $c_1$
\begin{align}
R(\alpha_1 \mid x = 3) &= 0   + 0.3 \cdot 0 = 0 \\
R(\alpha_2 \mid x = 3) &= 0.2 \cdot \frac{1}{30} + 0 = \frac{7}{30} \\
\end{align}

Classify $x_2$ as $c_2$
\begin{align}
R(\alpha_1 \mid x = 7) &= 0   + 0.3 \cdot \frac{1}{30} = \frac{1}{100}\\
R(\alpha_2 \mid x = 7) &= 0.2 \cdot \frac{1}{30} + 0 = \frac{1}{150} \\
\end{align}

Classify $x_3$ as $c_2$
\begin{align}
R(\alpha_1 \mid x = 8) &= 0   + 0.3 \cdot \frac{1}{15} = \frac{1}{50}\\
R(\alpha_2 \mid x = 8) &= 0.2 \cdot \frac{1}{60} + 0 = \frac{1}{300} \\
\end{align}

Classify $x_4$ as $c_2$
\begin{align}
R(\alpha_1 \mid x = 9) &= 0   + 0.3 \cdot \frac{2}{15} = \frac{1}{15}\\
R(\alpha_2 \mid x = 9) &= 0.2 \cdot 0 + 0 = 0 \\
\end{align}

---
## Exercise 4, Question 2


<img src="figures/exam-jun-2018/ex4.2.png" width="800" />





---
## Exercise 4, Question 2 a)


<img src="figures/exam-jun-2018/ex4.2a.png" width="800" />





In order to classify the test samples using a linear classifier, we can augment the data representations using a mapping that transforms the original samples to another feature space where the samples become linearly separable.

---
## Exercise 4, Question 2 b)


<img src="figures/exam-jun-2018/ex4.2b.png" width="800" />





We can use the second order polynomial kernel function:

$$
\phi(\mathbf{x}) = 
\begin{bmatrix}
x_1^2 \\
\sqrt{2} x_1 x_2 \\
x_2^2 
\end{bmatrix}
$$

---
## Exercise 4, Question 2 c)


<img src="figures/exam-jun-2018/ex4.2c.png" width="800" />







<img src="figures/exam-jun-2018/sol-ex4.2.b.png" width="200" />




---
## Exercise 4, Question 3 a)


<img src="figures/exam-jun-2018/ex4.3a.png" width="800" />





---
## Exercise 4, Question 3 b)


<img src="figures/exam-jun-2018/ex4.3b.png" width="800" />





---
## Exercise 4, Question 3 c)


<img src="figures/exam-jun-2018/ex4.3c.png" width="800" />





---
## Exercise 4, Question 3 d)


<img src="figures/exam-jun-2018/ex4.3d.png" width="800" />



